File: clojure/bird-watcher/src/bird_watcher.clj
(ns bird-watcher)

(def last-week
  [0 2 5 3 7 8 4]
  )

(defn today [birds]
  (get birds 6)
  )

(defn inc-bird [birds]
  (assoc birds 6 (+ (today birds) 1))
  )

(defn day-without-birds? [birds]
  (not (every? (fn [x] (> x 0)) birds))
  )

(defn n-days-count [birds n]
  (reduce + (subvec birds 0 n))
  )

(defn busy-days [birds]
  (count (filter (fn [x] (> x 4)) birds))
  )

(defn odd-week? [birds]
  (every? (fn [x] (or (= x 1) (= x 0))) birds)
  )

File: javascript/bird-watcher/bird-watcher.js
// @ts-check
//
// The line above enables type checking for this file. Various IDEs interpret
// the @ts-check directive. It will give you helpful autocompletion when
// implementing this exercise.

/**
 * Calculates the total bird count.
 *
 * @param {number[]} birdsPerDay
 * @returns {number} total bird count
 */
export function totalBirdCount(birdsPerDay) {
    return birdsPerDay.reduce((a, b) => a + b, 0);
}

/**
 * Calculates the total number of birds seen in a specific week.
 *
 * @param {number[]} birdsPerDay
 * @param {number} week
 * @returns {number} birds counted in the given week
 */
export function birdsInWeek(birdsPerDay, week) {
    let startIndex = (week - 1) * 7;

    return birdsPerDay.slice(startIndex, startIndex + 7)
                .reduce((a, b) => a + b, 0);
}

/**
 * Fixes the counting mistake by increasing the bird count
 * by one for every second day.
 *
 * @param {number[]} birdsPerDay
 * @returns {number[]} corrected bird count data
 */
export function fixBirdCountLog(birdsPerDay) {

    for (let index = 0; index < birdsPerDay.length; index++) {
        let num = birdsPerDay[index];
        birdsPerDay[index] = index % 2 == 0 ? num + 1 : num
    }

    return birdsPerDay;
}
